src/main/kotlin/tech/relaycorp/relaynet/cogrpc/client/CogRPCClient.kt

```diff
@@ -66,39 +66,46 @@ private constructor(
     fun deliverCargo(cargoes: Iterable<CargoDeliveryRequest>): Flow<String> {
         if (cargoes.none()) return emptyFlow()
 
+        var deliveryObserver: StreamObserver<CargoDelivery>? = null
         val cargoesToAck = mutableListOf<String>()
         val ackChannel = BroadcastChannel<String>(1)
         val ackObserver = object : StreamObserver<CargoDeliveryAck> {
             override fun onNext(value: CargoDeliveryAck) {
                 logger.info("deliverCargo ack ${value.id}")
                 ackChannel.sendBlocking(value.id)
                 cargoesToAck.remove(value.id)
+                if (cargoesToAck.isEmpty()) {
+                    logger.info("deliverCargo complete")
+                    deliveryObserver?.onCompleted()
+                }
             }
 
             override fun onError(t: Throwable) {
                 logger.log(Level.WARNING, "deliverCargo ack error", t)
                 ackChannel.close(CogRPCException(t))
+                logger.info("deliverCargo complete")
```

Comment:
Same as above: I'd suggest to combine these logs into one warning so they don't like separate events. E.g., `logger.log(Level.WARNING, "Ending deliverCargo due to ack error", t)`.